Panic during shutdown of server

During a shutdown using the `inmem` index:

```
^C[I] 2017-03-20T20:09:19Z Signal received, initializing clean shutdown...
[I] 2017-03-20T20:09:19Z Waiting for clean shutdown...
[I] 2017-03-20T20:09:19Z snapshot listener closed service=snapshot
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xe8 pc=0x82528f]

goroutine 77 [running]:
github.com/influxdata/influxdb/tsdb.(*Shard).SeriesSketches(0xc42016a9c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/edd/Development/golang/src/github.com/influxdata/influxdb/tsdb/shard.go:495 +0xaf
github.com/influxdata/influxdb/tsdb.(*Store).SeriesCardinality.func1(0xc42016a9c0, 0xc44c74af20, 0xc44c738140, 0x1, 0x1, 0xc44c736ff0, 0x0)
	/Users/edd/Development/golang/src/github.com/influxdata/influxdb/tsdb/store.go:717 +0xcc
github.com/influxdata/influxdb/tsdb.(*Store).estimateCardinality(0xc42016e280, 0xc4202c040c, 0x6, 0xc372a8, 0xf6e910, 0xc4200fa1d8, 0xc44c736ff0)
	/Users/edd/Development/golang/src/github.com/influxdata/influxdb/tsdb/store.go:691 +0x101
github.com/influxdata/influxdb/tsdb.(*Store).SeriesCardinality(0xc42016e280, 0xc4202c040c, 0x6, 0xc44c746ab8, 0xc44c73aed0, 0x0)
	/Users/edd/Development/golang/src/github.com/influxdata/influxdb/tsdb/store.go:718 +0x4b
github.com/influxdata/influxdb/tsdb.(*Store).Statistics(0xc42016e280, 0xc420261200, 0x1, 0x1, 0xc44c74ad20)
	/Users/edd/Development/golang/src/github.com/influxdata/influxdb/tsdb/store.go:95 +0x13e
github.com/influxdata/influxdb/cmd/influxd/run.(*Server).Statistics(0xc42016e140, 0xc420261200, 0xc420180140, 0xc4200fa560, 0x440b00)
	/Users/edd/Development/golang/src/github.com/influxdata/influxdb/cmd/influxd/run/server.go:216 +0xeb
github.com/influxdata/influxdb/monitor.(*Monitor).gatherStatistics(0xc4201800f0, 0xc44c738128, 0x1, 0x1, 0xc420261200, 0x0, 0x0, 0x0)
	/Users/edd/Development/golang/src/github.com/influxdata/influxdb/monitor/service.go:318 +0xd3
github.com/influxdata/influxdb/monitor.(*Monitor).Statistics(0xc4201800f0, 0xc420261200, 0xf76e80, 0x2540be400, 0xed0622da0, 0xc400000000, 0xf76e80)
	/Users/edd/Development/golang/src/github.com/influxdata/influxdb/monitor/service.go:310 +0xe65
github.com/influxdata/influxdb/monitor.(*Monitor).storeStatistics(0xc4201800f0)
	/Users/edd/Development/golang/src/github.com/influxdata/influxdb/monitor/service.go:411 +0x401
created by github.com/influxdata/influxdb/monitor.(*Monitor).Open
	/Users/edd/Development/golang/src/github.com/influxdata/influxdb/monitor/service.go:122 +0x2f8
ubuntu@ip-172-30-2-55:~$
```

--- a/circle.yml
+++ b/circle.yml
@@ -22,6 +22,8 @@ test:
     override:
         - bash circle-test.sh:
             parallel: true
+            # Race tests using 960s timeout
+            timeout: 960
 
 deployment:
   release:

--- a/pkg/estimator/hll/hll.go
+++ b/pkg/estimator/hll/hll.go
@@ -174,6 +174,11 @@ func (h *Plus) Count() uint64 {
 // If HyperLogLogPlus h is using the sparse representation, it will be converted
 // to the normal representation.
 func (h *Plus) Merge(s estimator.Sketch) error {
+	if s == nil {
+		// Nothing to do
+		return nil
+	}
+
 	other, ok := s.(*Plus)
 	if !ok {
 		return fmt.Errorf("wrong type for merging: %T", other)

--- a/tsdb/engine/tsm1/cache_test.go
+++ b/tsdb/engine/tsm1/cache_test.go
@@ -519,6 +519,10 @@ func TestCache_CacheWriteMemoryExceeded(t *testing.T) {
 }
 
 func TestCache_Deduplicate_Concurrent(t *testing.T) {
+	if testing.Short() || os.Getenv("GORACE") != "" {
+		t.Skip("Skipping test in short mode.")
+	}
+
 	values := make(map[string][]Value)
 
 	for i := 0; i < 1000; i++ {

--- a/tsdb/engine/tsm1/wal.go
+++ b/tsdb/engine/tsm1/wal.go
@@ -77,6 +77,10 @@ const (
 
 // WAL represents the write-ahead log used for writing TSM files.
 type WAL struct {
+	// goroutines waiting for the next fsync
+	syncCount   uint64
+	syncWaiters chan chan error
+
 	mu            sync.RWMutex
 	lastWriteTime time.Time
 
@@ -89,9 +93,6 @@ type WAL struct {
 	// cache and flush variables
 	once    sync.Once
 	closing chan struct{}
-	// goroutines waiting for the next fsync
-	syncWaiters chan chan error
-	syncCount   uint64
 
 	// syncDelay sets the duration to wait before fsyncing writes.  A value of 0 (default)
 	// will cause every write to be fsync'd.  This must be set before the WAL

--- a/tsdb/shard.go
+++ b/tsdb/shard.go
@@ -492,6 +492,9 @@ func (s *Shard) SeriesN() int64 {
 func (s *Shard) SeriesSketches() (estimator.Sketch, estimator.Sketch, error) {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
+	if s.engine == nil {
+		return nil, nil, nil
+	}
 	return s.engine.SeriesSketches()
 }
 
@@ -499,6 +502,9 @@ func (s *Shard) SeriesSketches() (estimator.Sketch, estimator.Sketch, error) {
 func (s *Shard) MeasurementsSketches() (estimator.Sketch, estimator.Sketch, error) {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
+	if s.engine == nil {
+		return nil, nil, nil
+	}
 	return s.engine.MeasurementsSketches()
 }
 

--- a/tsdb/store_test.go
+++ b/tsdb/store_test.go
@@ -25,6 +25,8 @@ import (
 // Ensure the store can delete a retention policy and all shards under
 // it.
 func TestStore_DeleteRetentionPolicy(t *testing.T) {
+	t.Parallel()
+
 	s := MustOpenStore()
 	defer s.Close()
 
@@ -93,6 +95,8 @@ func TestStore_DeleteRetentionPolicy(t *testing.T) {
 
 // Ensure the store can create a new shard.
 func TestStore_CreateShard(t *testing.T) {
+	t.Parallel()
+
 	s := MustOpenStore()
 	defer s.Close()
 
@@ -122,6 +126,8 @@ func TestStore_CreateShard(t *testing.T) {
 
 // Ensure the store can delete an existing shard.
 func TestStore_DeleteShard(t *testing.T) {
+	t.Parallel()
+
 	s := MustOpenStore()
 	defer s.Close()
 
@@ -142,6 +148,8 @@ func TestStore_DeleteShard(t *testing.T) {
 
 // Ensure the store can create a snapshot to a shard.
 func TestStore_CreateShardSnapShot(t *testing.T) {
+	t.Parallel()
+
 	s := MustOpenStore()
 	defer s.Close()
 
@@ -162,6 +170,8 @@ func TestStore_CreateShardSnapShot(t *testing.T) {
 }
 
 func TestStore_Open(t *testing.T) {
+	t.Parallel()
+
 	s := NewStore()
 	defer s.Close()
 
@@ -197,6 +207,8 @@ func TestStore_Open(t *testing.T) {
 
 // Ensure the store reports an error when it can't open a database directory.
 func TestStore_Open_InvalidDatabaseFile(t *testing.T) {
+	t.Parallel()
+
 	s := NewStore()
 	defer s.Close()
 
@@ -215,6 +227,8 @@ func TestStore_Open_InvalidDatabaseFile(t *testing.T) {
 
 // Ensure the store reports an error when it can't open a retention policy.
 func TestStore_Open_InvalidRetentionPolicy(t *testing.T) {
+	t.Parallel()
+
 	s := NewStore()
 	defer s.Close()
 
@@ -237,6 +251,8 @@ func TestStore_Open_InvalidRetentionPolicy(t *testing.T) {
 
 // Ensure the store reports an error when it can't open a retention policy.
 func TestStore_Open_InvalidShard(t *testing.T) {
+	t.Parallel()
+
 	s := NewStore()
 	defer s.Close()
 
@@ -259,6 +275,8 @@ func TestStore_Open_InvalidShard(t *testing.T) {
 
 // Ensure shards can create iterators.
 func TestShards_CreateIterator(t *testing.T) {
+	t.Parallel()
+
 	s := MustOpenStore()
 	defer s.Close()
 
@@ -334,6 +352,8 @@ func TestShards_CreateIterator(t *testing.T) {
 
 // Ensure the store can backup a shard and another store can restore it.
 func TestStore_BackupRestoreShard(t *testing.T) {
+	t.Parallel()
+
 	s0, s1 := MustOpenStore(), MustOpenStore()
 	defer s0.Close()
 	defer s1.Close()
@@ -462,6 +482,8 @@ func testStoreCardinalityTombstoning(t *testing.T, store *Store) {
 }
 
 func TestStore_Cardinality_Tombstoning_Inmem(t *testing.T) {
+	t.Parallel()
+
 	store := NewStore()
 	store.EngineOptions.Config.Index = "inmem"
 	if err := store.Open(); err != nil {
@@ -472,6 +494,8 @@ func TestStore_Cardinality_Tombstoning_Inmem(t *testing.T) {
 }
 
 func TestStore_Cardinality_Tombstoning_TSI(t *testing.T) {
+	t.Parallel()
+
 	store := NewStore()
 	store.EngineOptions.Config.Index = "tsi1"
 	if err := store.Open(); err != nil {
@@ -530,6 +554,8 @@ func testStoreCardinalityUnique(t *testing.T, store *Store) {
 }
 
 func TestStore_Cardinality_Unique_Inmem(t *testing.T) {
+	t.Parallel()
+
 	store := NewStore()
 	store.EngineOptions.Config.Index = "inmem"
 	if err := store.Open(); err != nil {
@@ -540,6 +566,8 @@ func TestStore_Cardinality_Unique_Inmem(t *testing.T) {
 }
 
 func TestStore_Cardinality_Unique_TSI1(t *testing.T) {
+	t.Parallel()
+
 	store := NewStore()
 	store.EngineOptions.Config.Index = "tsi1"
 	if err := store.Open(); err != nil {
@@ -614,6 +642,8 @@ func testStoreCardinalityDuplicates(t *testing.T, store *Store) {
 }
 
 func TestStore_Cardinality_Duplicates_Inmem(t *testing.T) {
+	t.Parallel()
+
 	store := NewStore()
 	store.EngineOptions.Config.Index = "inmem"
 	if err := store.Open(); err != nil {
@@ -624,6 +654,8 @@ func TestStore_Cardinality_Duplicates_Inmem(t *testing.T) {
 }
 
 func TestStore_Cardinality_Duplicates_TSI1(t *testing.T) {
+	t.Parallel()
+
 	store := NewStore()
 	store.EngineOptions.Config.Index = "tsi1"
 	if err := store.Open(); err != nil {
@@ -684,6 +716,8 @@ func testStoreCardinalityCompactions(t *testing.T, store *Store) {
 }
 
 func TestStore_Cardinality_Compactions_Inmem(t *testing.T) {
+	t.Parallel()
+
 	store := NewStore()
 	store.EngineOptions.Config.Index = "inmem"
 	if err := store.Open(); err != nil {
@@ -694,6 +728,8 @@ func TestStore_Cardinality_Compactions_Inmem(t *testing.T) {
 }
 
 func TestStore_Cardinality_Compactions_TSI1(t *testing.T) {
+	t.Parallel()
+
 	store := NewStore()
 	store.EngineOptions.Config.Index = "tsi1"
 	if err := store.Open(); err != nil {

--- a/tsdb/store_test.go
+++ b/tsdb/store_test.go
@@ -420,7 +420,7 @@ func TestStore_BackupRestoreShard(t *testing.T) {
 }
 
 func testStoreCardinalityTombstoning(t *testing.T, store *Store) {
-	if testing.Short() {
+	if testing.Short() || os.Getenv("GORACE") != "" {
 		t.Skip("Skipping test in short mode.")
 	}
 
@@ -506,7 +506,7 @@ func TestStore_Cardinality_Tombstoning_TSI(t *testing.T) {
 }
 
 func testStoreCardinalityUnique(t *testing.T, store *Store) {
-	if testing.Short() {
+	if testing.Short() || os.Getenv("GORACE") != "" {
 		t.Skip("Skipping test in short mode.")
 	}
 
@@ -580,7 +580,7 @@ func TestStore_Cardinality_Unique_TSI1(t *testing.T) {
 // This test tests cardinality estimation when series data is duplicated across
 // multiple shards.
 func testStoreCardinalityDuplicates(t *testing.T, store *Store) {
-	if testing.Short() {
+	if testing.Short() || os.Getenv("GORACE") != "" {
 		t.Skip("Skipping test in short mode.")
 	}
 
@@ -668,7 +668,7 @@ func TestStore_Cardinality_Duplicates_TSI1(t *testing.T) {
 // Creates a large number of series in multiple shards, which will force
 // compactions to occur.
 func testStoreCardinalityCompactions(t *testing.T, store *Store) {
-	if testing.Short() {
+	if testing.Short() || os.Getenv("GORACE") != "" {
 		t.Skip("Skipping test in short mode.")
 	}
 

